src/Plugin/Commerce/PaymentGateway/Midtrans.php

```diff
@@ -128,65 +150,120 @@ public function submitConfigurationForm(array &$form, FormStateInterface $form_s
+
+    $this->messenger()->addMessage($this->t($message));
+  }
+
+  protected function detailPaymentType($data) {
+    $result = '-';
+    if ($data->payment_type == 'credit_card') {
+      $result =  ucwords($data->card_type).' Card - Mask Card: '.$data->masked_card;
     }
+    else if ($data->payment_type == 'qris') {
+      $issuer = isset($data->issuer) ? ' - Issuer: '.ucwords($data->issuer) : '';
+      $result = 'QRIS - Acquirer: '.ucwords($data->acquirer).$issuer;
+    } else if ($data->payment_type == 'cstore') {
+      $result = ucwords($data->store).' - Payment Code: '.$data->payment_code;
+    }
+    else if ($data->payment_type == 'echannel') {
+      $result = 'Mandiri Bill - Bill Number: '.$data->bill_key;
+    }
+    else if ($data->payment_type == 'bank_transfer') {
+      if (isset($data->permata_va_number)){
+        $result = 'Permata VA - '.$data->permata_va_number;
+      }
+      else {
+        $result = strtoupper($data->va_numbers[0]->bank).' VA - '.$data->va_numbers[0]->va_number;
+      }
+    }
+    else {
+      $result = ucwords($data->payment_type);
+    }
+
+    return $result;
   }
 
   /**
    * {@inheritdoc}
    */
   public function onNotify(Request $request) {
-    \Midtrans\Config::$serverKey =  $this->getConfiguration()['server_key'];
-    \Midtrans\Config::$isProduction = ($this->getMode() == 'production') ? TRUE : FALSE;
-    $response = new \Midtrans\Notification();
-    /** @var \Drupal\commerce_payment\PaymentStorage $payment_storage */
-    $payment_storage = $this->entityTypeManager->getStorage('commerce_payment');
-    /** @var \Drupal\commerce_payment\Entity\Payment $payment */
-    //$payment = $payment_storage->loadByRemoteId($response->order_id);
-    /** @var \Drupal\commerce_order\Entity\Order $order */
-    //$order = $payment->getOrder();
-    
-    //error_log('Response from Midtrans : '. print_r($response, TRUE)); //debugan  
+    $this->midtransConfig();
+    $notification = new \Midtrans\Notification();
+    $response = $notification->getResponse();
     $payment = $this->loadPaymentByOrderId($response->order_id);
+    $order = $payment->getOrder();
+
+    // when use snap popup, some payment methods like akulaku or direct debit, will redirect to 3rd party website
+    // and never returns to the site, causes the order still unplaced (draft) and stuck in checkout
+    // and make sure that the order is placed when got status pending (order was created in midtrans)
+    if ($order->getState()->getId() == 'draft' && $response->transaction_status == 'pending') {
+      $order_state = $order->getState();
+      $order_state->applyTransitionById('place');
+      $order->unlock();
+    }
+
+    // change payment remote id with transaction id
+    $payment->setRemoteId($response->transaction_id);
+    $payment->save();
+
+    $message = 'orderID '.$response->order_id.' - '.$response->payment_type.' - '.$response->transaction_status;
+    \Drupal::logger('commerce_midtrans')->info($message);
```

Comment:
what is the default state of the logging? is it always logging these messages? because merchant may feel this log can spam their storage space if it always on.